dead_letter system table to keep erroneous incoming messages

ilejn opened the pull request:
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a `system.dead_letter_queue` table to keep erroneous incoming messages from engines like Kafka

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

Implements one of requests from  https://github.com/ClickHouse/ClickHouse/issues/65297#issue-2353551382

ilejn committed:
kafka_dead_letter_queue: initial, legacy tests work

ilejn committed:
kafka_dead_letter_queue: test passed

ilejn committed:
kafka_dead_letter_queue: stream type

ilejn committed:
kafka_dead_letter_queue: rename, small cleanup

ilejn committed:
kafka_dead_letter_queue: StorageKafka2 covered

ilejn committed:
kafka_dead_letter_queue: docs

ilejn committed:
kafka_dead_letter_queue: debug logging removed

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #68873 from ilejn/ClickHouse:kafka_dead_letter_queue -->
*This is an automated comment for commit e411d7f50fdf8920d25e95daaab65400451165e6 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68873/e411d7f50fdf8920d25e95daaab65400451165e6/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/68873/e411d7f50fdf8920d25e95daaab65400451165e6/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

ilejn committed:
kafka_dead_letter_queue: black formatter

ilejn committed:
kafka_dead_letter_queue: Nullable, forgotten doc added

ilejn committed:
kafka_dead_letter_queue: RabbitMQ initial

ilejn committed:
kafka_dead_letter_queue: Details struct introduced

ilejn committed:
kafka_dead_letter_queue: std::variant and visitor

ilejn committed:
kafka_dead_letter_queue: Kafka and RabbitMQ work

ilejn committed:
kafka_dead_letter_queue: style fixes

ilejn committed:
kafka_dead_letter_queue: forgotten config

ilejn committed:
kafka_dead_letter_queue: prefixes, no Nullable

ilejn committed:
kafka_dead_letter_queue: topics are unique (for flaky test)

ilejn committed:
kafka_dead_letter_queue: minor fixes, clang-tidy

ilejn committed:
kafka_dead_letter_queue: topics are even more unique

ilejn commented:
LDAP issue is addressed by https://github.com/ClickHouse/ClickHouse/pull/70755

ilejn committed:
kafka_dead_letter_queue: compiled with new master, cleanup

ilejn committed:
kafka_dead_letter_queue: tiny cleanup

ilejn committed:
kafka_dead_letter_queue: make black formatter happy

ilejn commented:
Test failures are most likely not not related.
Looking into test_storage_rabbitmq/test.py::test_rabbitmq_restore_failed_connection_without_losses_

antaljanosbenjamin commented on docs/en/operations/system-tables/dead_letter_queue.md:
```diff
@@ -0,0 +1,71 @@
+---
+# dead_letter_queue
+
+Contains information about messages received via a stream engine and parsed with an errors. Currently implemented for Kafka and RabbitMQ.
```
Maybe `streaming engine` is also okay.
```suggestion
Contains information about messages received via a stream like engine and parsed with an errors. Currently implemented for Kafka and RabbitMQ.
```

ilejn replied:
Thanks. I think I like "streaming engine" more.

antaljanosbenjamin commented on docs/en/operations/system-tables/dead_letter_queue.md:
```diff
@@ -0,0 +1,71 @@
+
+Contains information about messages received via a stream engine and parsed with an errors. Currently implemented for Kafka and RabbitMQ.
+
+Logging is controlled by `dead_letter_queue` of `kafka_handle_error_mode` setting.
```
```suggestion
Logging is enabled by specifying `dead_letter_queue` for the engine specific `handle_error_mode` setting.
```

antaljanosbenjamin commented on docs/en/operations/system-tables/dead_letter_queue.md:
```diff
@@ -0,0 +1,71 @@
+
+Columns:
+
+- `stream_type` ([Enum8](../../sql-reference/data-types/enum.md)) - Stream type. Possible values: 'Kafka'.
```
```suggestion
- `stream_type` ([Enum8](../../sql-reference/data-types/enum.md)) - Stream type. Possible values: 'Kafka' or 'RabbitMQ'.
```

ilejn replied:
In similar case in settings.md it is "and", not "or".
ChatGPT agrees that "and" is slightly better.

antaljanosbenjamin replied:
Yeah, that is completely fine for me as long as it is mentioned. Thanks for checking it though!

antaljanosbenjamin commented on docs/en/operations/system-tables/dead_letter_queue.md:
```diff
@@ -0,0 +1,71 @@
+- `event_date` ([Date](../../sql-reference/data-types/date.md)) - Message consuming date.
+- `event_time` ([DateTime](../../sql-reference/data-types/datetime.md)) - Message consuming date and time.
+- `event_time_microseconds` ([DateTime64](../../sql-reference/data-types/datetime64.md)) - Message consuming time with microseconds precision.
+- `database_name` ([LowCardinality(String)](../../sql-reference/data-types/string.md)) - ClickHouse database Kafka table belongs to.
```
```suggestion
- `database_name` ([LowCardinality(String)](../../sql-reference/data-types/string.md)) - ClickHouse database the table belongs to.
```

antaljanosbenjamin commented on src/Core/StreamingHandleErrorMode.h:
```diff
@@ -9,7 +9,13 @@ enum class StreamingHandleErrorMode : uint8_t
-    /*FIXED_SYSTEM_TABLE, Put errors to in a fixed system table likely system.kafka_errors. This is not implemented now.  */
+};
+
+enum class ExtStreamingHandleErrorMode : uint8_t
+{
+    DEFAULT = 0, // Ignore errors with threshold.
+    STREAM, // Put errors to stream in the virtual column named ``_error.
+    DEAD_LETTER_QUEUE
     /*CUSTOM_SYSTEM_TABLE, Put errors to in a custom system table. This is not implemented now.  */
 };
```
Have you thought about just extending `StreamingHandleErrorMode` and throw an error when `dead_letter_queue` is specified for an engine that doesn't support that? For me that feels like a better approach, because the two enums are really about the same setting.

ilejn replied:
Honestly I don't think that it is a right thing to do, basically it compromises the idea of typed enums. As a practical reason - if we use extended enum everywhere we will  struggle with compiler warnings.
What we need is inheritance, but it is not possible out of the box.

antaljanosbenjamin replied:
Ahh, okay. I thought that we can set the default error handling mode in the config, but after checking it I realized it is not possible. I think this should fine then.

antaljanosbenjamin replied:
I couldn't stop thinking about this and asked some of the collegues. In the end I think we should unify these enums. My reasons:
1. What if somebody wants to add another way of handling errors is stream like engine? Depending on whether it is relevant to Kafka/RabbitMQ/other engines, we might want to make new enums. This doesn't feel good. If it affects all engines, then they have to extend two enums.
2. Similar to the previous one, but slightly different: if somebody wants to change what the `DEFAULT` value mean, they might miss to change the ones with the extended enum.

Therefore I think the better approach is to have a single enum and make sure that `DEAD_LETTER_QUEUE` is handled in every engines either by using the new system table or throwing an exception on table creation (similarly to other, invalid setting values).

ilejn replied:
@antaljanosbenjamin , could you elaborate, please.

If you believe that any fancy and engine-specific techniques are to be handled individually and should not be included in a global enum, I agree.

We assume that we understand nature of streaming tables and provide some unification. Moreover, we have cross-engine dead_letter_queue. That's why I believe that introducing ExtStreamingHandleErrorMode is reasonable, risk that it would be Ext2StreamingHandleErrorMode whatever is minimal. Although it is too early to get rid of StreamingHandleErrorMode completely.

If you believe that we have to (eventually) support dead_letter_queue and switch to ExtStreamingHandleErrorMode for all streaming engine, I agree.

If you believe that we have to switch to ExtStreamingHandleErrorMode for all streaming engine even if a particular streaming engine does not support dead_letter_queue, I would NOT agree. What is the point?

As you may remember, initially dead_letter_queue was requested for kafka. I've thought that it would be nice to have it in a more generic way, while I don't have enough courage to handle engines like NATS I know nothing about.


If you do believe that it is a good idea to use ExtStreamingHandleErrorMode for all streaming engines right now despite my considerations above - let me know, I'll implement this, it's not a big deal.

antaljanosbenjamin replied:
> If you believe that we have to (eventually) support dead_letter_queue and switch to ExtStreamingHandleErrorMode for all streaming engine, I agree.

Agree. We don't have to support that now for all streaming engines.

> If you believe that we have to switch to ExtStreamingHandleErrorMode for all streaming engine even if a particular streaming engine does not support dead_letter_queue, I would NOT agree. What is the point?

The point is not to split streaming engine settings. In ClickHouse I think there quite some things that are handled in a way that some engines throws if they don't support it. Not the best, but an example is StorageKafka(2) [only supports](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/Kafka/StorageKafkaUtils.cpp#L225-L232) ordinary columns but not `DEFAULT/MATERIALIZED/EPHEMERAL` columns. It is not the best example, because it depends on our grammar, not a single enum, but it proves that you can easily specify a column/setting for an engine which it cannot handle. In my opinion having an extra check in all streaming engines that doesn't support dead letter queue is much better solution than having two enums that are almost the same. 

> As you may remember, initially dead_letter_queue was requested for kafka. I've thought that it would be nice to have it in a more generic way, while I don't have enough courage to handle engines like NATS I know nothing about.

And that is perfectly fine. I think you have already over delivered it by implementing it to RabbitMQ, I and I think the whole community appreciates it. 

> If you genuinely believe that implementing ExtStreamingHandleErrorMode across all streaming engines is a good idea despite my concerns, please let me know, and I'll take care of it - it's not a major undertaking.
Probably we can introduce error handling traits on a per-engine basis.

Just to be clear: we don't have to support dead letter queue in all streaming engines, but we should use the same enum for error handling mode across all streaming engines and just throw an exception when the engine doesn't support it.

How does this sound? If you are still not convinced that unifying these enums is better, then feel to say your concerns or point out if I haven't addressed your already mentioned concerns properly. I would like to reach a compromise here, because I truly beleive that lack of compromise means some of the parties (that can easily be me) are missing something important.

antaljanosbenjamin commented on src/Interpreters/DeadLetterQueue.h:
```diff
@@ -0,0 +1,59 @@
+    String error;
+
+    /// Engine specific details
+    struct KafkaDetails
+    {
+        String topic_name;
+        Int64 partition;
+        Int64 offset;
+    };
+    struct RabbitMQDetails
```
Maybe also include the message key, which is in the `_key` virtual column is storage kafka.

ilejn replied:
Because of an unclear reason _key is not populated in case of an error.

antaljanosbenjamin commented on src/Storages/Kafka/KafkaSource.cpp:
```diff
@@ -118,28 +117,32 @@ Chunk KafkaSource::generateImpl()
-                // We could already push some rows to result_columns before exception, we need to fix it.
-                result_columns[i]->rollback(*checkpoints[i]);
+            case ExtStreamingHandleErrorMode::STREAM:
+            case ExtStreamingHandleErrorMode::DEAD_LETTER_QUEUE: {
```
```suggestion
            case ExtStreamingHandleErrorMode::DEAD_LETTER_QUEUE:
            {
```

antaljanosbenjamin commented on src/Storages/Kafka/KafkaSource.cpp:
```diff
@@ -118,28 +117,32 @@ Chunk KafkaSource::generateImpl()
+                }
+                break;
+            }
+            case ExtStreamingHandleErrorMode::DEFAULT: {
```
```suggestion
            case ExtStreamingHandleErrorMode::DEFAULT:
            {
```

antaljanosbenjamin commented on src/Storages/RabbitMQ/RabbitMQSource.cpp:
```diff
@@ -167,22 +171,26 @@ Chunk RabbitMQSource::generateImpl()
-                // We could already push some rows to result_columns before exception, we need to fix it.
-                result_columns[i]->rollback(*checkpoints[i]);
+            case ExtStreamingHandleErrorMode::STREAM:
+            case ExtStreamingHandleErrorMode::DEAD_LETTER_QUEUE: {
```
```suggestion
            case ExtStreamingHandleErrorMode::DEAD_LETTER_QUEUE:
            {
```

antaljanosbenjamin commented on src/Storages/Kafka/StorageKafka2.cpp:
```diff
@@ -967,6 +974,32 @@ StorageKafka2::PolledBatchInfo StorageKafka2::pollConsumer(
+                {
+                    if (exception_message)
+                    {
+
+                        const auto time_now = std::chrono::system_clock::now();
+
+                        auto dead_letter_queue = getContext()->getDeadLetterQueue();
+                        dead_letter_queue->add(
+                            DeadLetterQueueElement{
+                                .stream_type = DeadLetterQueueElement::StreamType::Kafka,
+                                .event_time = timeInSeconds(time_now),
+                                .event_time_microseconds = timeInMicroseconds(time_now),
+                                .database_name = getStorageID().database_name,
+                                .table_name = getStorageID().table_name,
+                                .raw_message = consumer.currentPayload(),
+                                .error = exception_message.value(),
+                                .details = DeadLetterQueueElement::KafkaDetails{
+                                    .topic_name = consumer.currentTopic(),
+                                    .partition = consumer.currentPartition(),
+                                    .offset = consumer.currentPartition()
+                                }
+                            });
```
```suggestion
                        const auto time_now = std::chrono::system_clock::now();
                        auto storage_id = storage.getStorageID();

                        auto dead_letter_queue = getContext()->getDeadLetterQueue();
                        dead_letter_queue->add(
                            DeadLetterQueueElement{
                                .stream_type = DeadLetterQueueElement::StreamType::Kafka,
                                .event_time = timeInSeconds(time_now),
                                .event_time_microseconds = timeInMicroseconds(time_now),
                                .database_name = storage_id.database_name,
                                .table_name = storage_id.table_name,
                                .raw_message = consumer.currentPayload(),
                                .error = exception_message.value(),
                                .details = DeadLetterQueueElement::KafkaDetails{
                                    .topic_name = consumer.currentTopic(),
                                    .partition = consumer.currentPartition(),
                                    .offset = consumer.currentPartition()
                                }
                            });
```

antaljanosbenjamin reviewed (commented):
In general the PR looks good, there are small stuff here and there, but nothing too serious. Thanks for the nice contribution again!

ilejn committed:
kafka_dead_letter_queue: fix rabbit flaky tests

ilejn committed:
kafka_dead_letter_queue: code review suggestions with _key

ilejn committed:
kafka_dead_letter_queue: remove kafka_key, minor improvements

ilejn committed:
Revert "kafka_dead_letter_queue: remove kafka_key, minor improvements"

This reverts commit fd73957cbafc5894accf0878ee67c087a91b5286.

ilejn committed:
kafka_dead_letter_queue: minor improvements

ilejn committed:
kafka_dead_letter_queue: make black happy

ilejn committed:
kafka_dead_letter_queue: no ExtStreamingHandleErrorMode

ilejn commented:
Hello @antaljanosbenjamin , any thoughts why https://github.com/ilejn/ClickHouse/commit/e6ab139cee5d51dbef4f7a490424fc776819c153 is ignored by CI ?

ilejn committed:
kafka_dead_letter_queue: test for unsupported

ilejn committed:
kafka_dead_letter_queue: no NATS test

ilejn commented:
Hello @antaljanosbenjamin , have a look please.

I am trying to fix test_storage_rabbitmq/test.py::test_rabbitmq_restore_failed_connection_without_losses_1 .

Note, that natural effect of reusing enum is 
```
CREATE TABLE queue
(
    `key` UInt64,
    `value` UInt64
)
ENGINE = NATS
SETTINGS nats_url = 'localhost:4444', nats_subjects = 'subject1,subject2', nats_format = 'JSONEachRow', nats_handle_error_mode = 'qwe'

Query id: 1be264da-ba7d-49b6-af0f-39aee139a8cf


Elapsed: 0.012 sec. 

Received exception from server (version 24.12.1):
Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Unexpected value of StreamingHandleErrorMode: 'qwe'. Must be one of ['dead_letter_queue', 'stream', 'default']. (BAD_ARGUMENTS)
```
(dead_letter_queue listed as an option).

antaljanosbenjamin commented:
> Note, that natural effect of reusing enum is

I think that is fine. If somebody will specify `dead_letter_queue`, then they will get to know that filelog/nats doesn't support that.

ilejn commented:
>  once the tests are green, we can merge this.

I've fixed two rabbit tests that were flaky.

No ideas about test_storage_rabbitmq/test.py::test_rabbitmq_restore_failed_connection_without_losses_1 yet.

antaljanosbenjamin commented:
We are missing `integration_run_flaky0_0.tar.zst` from the test report, so it is kind of hard to figure this out. As those are the only failures, I will start to restart the tests. If they pass, I can merge the PR.

In parallel I will try to ask people to check out why the tar is not in the report.

ilejn commented:
> We are missing `integration_run_flaky0_0.tar.zst` from the test report, so it is kind of hard to figure this out. As those are the only failures, I will start to restart the tests. If they pass, I can merge the PR.
> 
> In parallel I will try to ask people to check out why the tar is not in the report.

Why?
integration_run_flaky0_0.tar.zst exists in the bundle, but it does not help much.

What I can see is at the point of 83100 messages 
```
SELECT count(DISTINCT key) FROM test.view
```
stopped seeing new records (executeQuery: Read 83100 rows), 
and in logs we have
```
2024.12.11 01:31:25.598141 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83098, previous delivery tag: 83097, redelivered: false, failed delivery tags by this moment: 0, exception message: None
2024.12.11 01:31:25.598176 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83099, previous delivery tag: 83098, redelivered: false, failed delivery tags by this moment: 0, exception message: None
2024.12.11 01:31:25.598211 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83100, previous delivery tag: 83099, redelivered: false, failed delivery tags by this moment: 0, exception message: None
2024.12.11 01:31:25.598226 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Flushing 100 rows (max block size: 100, time: 11 / 1000 ms)
2024.12.11 01:31:25.598374 [ 548 ] {} <Trace> StorageRabbitMQ (rabbitmq_consume): RabbitMQSource is stopped (is_finished: true, consumer_stopped: 0)

```

ilejn commented:
after restart or something consuming goes on according to logs
```
2024.12.11 01:31:25.648810 [ 548 ] {} <Trace> StorageRabbitMQ (rabbitmq_consume): Consumer committed messages with deliveryTags up to 83300 on channel 1_0_aaramlasharipjrqppxlcxkktpvwihlo
2024.12.11 01:31:25.648845 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Number of attached views 1 for test.rabbitmq_consume (ce32f795-7ccd-41c4-91d0-fc26c2f353ac)
2024.12.11 01:31:25.648871 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Will start background loop to let messages be pushed to channel
2024.12.11 01:31:25.648934 [ 557 ] {} <Debug> StorageRabbitMQ (rabbitmq_consume): Background loop started
2024.12.11 01:31:25.649017 [ 597 ] {5de64bed-6c86-4fc8-9a84-b31edeb447be::all_1_828_376} <Debug> MergeTask::PrepareStage: Merging 2 parts: from all_1_827_375 to all_828_828_0 into Compact with storage Full
2024.12.11 01:31:25.649151 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Number of attached views 1 for test.rabbitmq_consume (ce32f795-7ccd-41c4-91d0-fc26c2f353ac)
2024.12.11 01:31:25.649174 [ 548 ] {} <Debug> StorageRabbitMQ (rabbitmq_consume): Started streaming to 1 attached views
2024.12.11 01:31:25.649179 [ 597 ] {5de64bed-6c86-4fc8-9a84-b31edeb447be::all_1_828_376} <Debug> MergeTask::PrepareStage: Selected MergeAlgorithm: Horizontal
2024.12.11 01:31:25.649354 [ 597 ] {5de64bed-6c86-4fc8-9a84-b31edeb447be::all_1_828_376} <Debug> MergeTreeSequentialSource: Reading 22 marks from part all_1_827_375, total 165400 rows starting from the beginning of the part
2024.12.11 01:31:25.649472 [ 597 ] {5de64bed-6c86-4fc8-9a84-b31edeb447be::all_1_828_376} <Debug> MergeTreeSequentialSource: Reading 2 marks from part all_828_828_0, total 200 rows starting from the beginning of the part
2024.12.11 01:31:25.649940 [ 548 ] {} <Trace> Planner: Query to stage Complete only analyze
2024.12.11 01:31:25.650085 [ 548 ] {} <Trace> Planner: Query from stage FetchColumns to stage Complete only analyze
2024.12.11 01:31:25.650311 [ 597 ] {5de64bed-6c86-4fc8-9a84-b31edeb447be::all_1_828_376} <Test> MergeTreeMarksLoader: Loading marks from path data.cmrk3
2024.12.11 01:31:25.650550 [ 548 ] {} <Test> Squashing: header columns 2
2024.12.11 01:31:25.651107 [ 548 ] {} <Test> InterpreterInsertQuery: Pipeline could use up to 1 thread
2024.12.11 01:31:25.651394 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83301, previous delivery tag: 0, redelivered: false, failed delivery tags by this moment: 0, exception message: None
2024.12.11 01:31:25.651405 [ 597 ] {5de64bed-6c86-4fc8-9a84-b31edeb447be::all_1_828_376} <Test> MergeTreeMarksLoader: Loading marks from path data.cmrk3
2024.12.11 01:31:25.651440 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83302, previous delivery tag: 83301, redelivered: false, failed delivery tags by this moment: 0, exception message: None
2024.12.11 01:31:25.651476 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83303, previous delivery tag: 83302, redelivered: false, failed delivery tags by this moment: 0, exception message: None
2024.12.11 01:31:25.651507 [ 548 ] {} <Test> StorageRabbitMQ (rabbitmq_consume): Pulled 1 rows, message delivery tag: 83304, previous delivery
```
but MV does not receive new records.

ilejn commented:
I have a suspicion, 
the test has check 
```
    while time.monotonic() < deadline:
        if int(instance.query("SELECT count() FROM test.view")) != 0:
            break
        time.sleep(0.1)
```
which effectively does not work in flaky check mode because we do not DROP test.view .

ilejn committed:
kafka_dead_letter_queue: maybe fix test_rabbitmq...without_losses_1

ilejn commented:
Hello @kssenii , could you reconfirm that check
```
    while int(instance.query('SELECT count() FROM test.view')) == 0:
        time.sleep(0.1)
```
in test_rabbitmq_restore_failed_connection_without_losses_1
makes sense in general?
If yes, 'drop test.view' should be helpful.

ilejn commented:
No luck ... test_storage_rabbitmq/test.py::test_rabbitmq_restore_failed_connection_without_losses_1 still flaky.

kssenii commented:
> No luck ... test_storage_rabbitmq/test.py::test_rabbitmq_restore_failed_connection_without_losses_1 still flaky.

btw it could have been flaky in the same way always, because recently there was a change in flaky check that it runs all repeated tests using the same clickhouse server, while previously it would start a new server for each test. E.g. this could have affected it, not the changes in your PR.

ilejn commented:
> > No luck ... test_storage_rabbitmq/test.py::test_rabbitmq_restore_failed_connection_without_losses_1 still flaky.
> 
> btw it could have been flaky in the same way always, because recently there was a change in flaky check that it runs all repeated tests using the same clickhouse server, while previously it would start a new server for each test. E.g. this could have affected it, not the changes in your PR.

Thanks, yes, it is more or less obvious that the PR does not cause the issue. 
Although the questions what is the actual problem and how to solve it remain open.

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: decrease flush_interval_ms to speedup test

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: merge collision fix

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

antaljanosbenjamin commented:
> Do you mind if I split kafka slow tests similarly to how it is done for s3_queue

The opposite, I would really appreciate that! Thanks in advance :) If you need help with anything, I am happy to help.

ilejn commented:
> The opposite, I would really appreciate that! Thanks in advance :) If you need help with anything, I am happy to help.

https://github.com/ClickHouse/ClickHouse/pull/77700 contains preparation steps,
I copied common stuff to integration/helpers.

ilejn committed:
kafka_dead_letter_queue: keep target table clean

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: black formatter

ilejn committed:
kafka_dead_letter_queue: RabbitMQ fix

ilejn committed:
kafka_dead_letter_queue: make clang_tidy happy

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: doc builder errors addressed

ilejn committed:
kafka_dead_letter_queue: rename dead_letter_queue to dead_letter

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: remove legacy

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: absolute path in docs

ilejn commented:
>  Do you think the PR is in a state where it makes sense to check it out?

@antaljanosbenjamin 
Thank you for the care.
I've quickly run through the changes, looks OK to me.
So the answer is 'Yes' assuming no related test failures.

ilejn committed:
kafka_dead_letter_queue: some columns renamed

ilejn committed:
kafka_dead_letter_queue: doc links fix

ilejn committed:
kafka_dead_letter_queue: doc links fix (2)

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

antaljanosbenjamin commented:
> Hm ... apparently I have to resolve a conflict.

Strange. Maybe it is a github issue? I checked the history of the file and it doesn't make sense in my opinion.

> Not sure what happened with build for fuzzers (CI issue?) .

Yes, we have issues with the fuzzers build, don't worry about that.

antaljanosbenjamin commented on src/Storages/Kafka/KafkaSource.cpp:
```diff
@@ -224,6 +243,26 @@ Chunk KafkaSource::generateImpl()
                     }
                 }
             }
+            if (is_dead_letter && exception_message)
```
Why is the `exception_message` part is important? I think when `dead_letter` is `true`, then `exception_message` is always set. Even if not, then we should still send the message to dead letter.
```suggestion
            if (is_dead_letter)
```

antaljanosbenjamin commented on src/Storages/RabbitMQ/StorageRabbitMQ.cpp:
```diff
@@ -8,6 +8,7 @@
 #include <Interpreters/DatabaseCatalog.h>
 #include <Interpreters/InterpreterInsertQuery.h>
 #include <Interpreters/InterpreterSelectQuery.h>
+#include <Interpreters/DeadLetter.h>
```
```suggestion
```

antaljanosbenjamin commented on src/Storages/RabbitMQ/RabbitMQSource.cpp:
```diff
@@ -249,6 +275,33 @@ Chunk RabbitMQSource::generateImpl()
                 }
             }
 
+            if (is_dead_letter && exception_message)
```
```suggestion
            if (is_dead_letter)
```

antaljanosbenjamin commented on src/Interpreters/DeadLetter.h:
```diff
@@ -0,0 +1,60 @@
+    std::variant<KafkaDetails, RabbitMQDetails> details;
+
+
+    static std::string name() { return "DeadLetter"; }
```
Can I ask why it was renamed to `DeadLetter` instead of `DeadLetterQueue`? `system.dead_letter` feels weird, `system.dead_letter_queue` feels more natural, but there might be something that I don't see or understand.

antaljanosbenjamin replied:
I see your point, especially the first one. It is not really a queue in that sense. Let's wait for @morsapaes with final verdict, but I can think of two more names:
1. `dead_letter*s*`: similar to `databases`, `tables`, `parts`
2. `dead_letter_messages`: it is kind of `dead_letter_mails`, but I think `messages` suits this use case more than `mails`.

antaljanosbenjamin commented on tests/integration/test_kafka_bad_messages/test.py:
```diff
@@ -20,7 +20,53 @@ def kafka_cluster():
-def test_bad_messages_parsing_stream(kafka_cluster):
+# check_method for bad_messages_parsing_mode
+def view_test(expected_num_messages, *_):
+    attempt = 0
+    rows = 0
+    while attempt < 500:
+        time.sleep(0.1)
+        rows = int(instance.query("SELECT count() FROM view"))
+        if rows == expected_num_messages:
+            break
+        attempt += 1
```
Can we use `instance.query_with_retry`?

antaljanosbenjamin commented on tests/integration/test_kafka_bad_messages/test.py:
```diff
@@ -20,7 +20,53 @@ def kafka_cluster():
+
+    attempt = 0
+    rows = 0
+    while attempt < 500:
+        time.sleep(0.1)
+        rows = int(
+            instance.query(
+                f"SELECT count() FROM system.dead_letter WHERE kafka_topic_name = '{topic_name}'"
+            )
+        )
+        if rows == expected_num_messages:
+            break
+        attempt += 1
+    assert rows == expected_num_messages
+
+    result = instance.query(
+        f"SELECT * FROM system.dead_letter WHERE kafka_topic_name = '{topic_name}' FORMAT Vertical"
+    )
+    logging.debug(f"system.dead_letter contains {result}")
```
Same here, `instance.query_with_retry`.

antaljanosbenjamin commented on tests/integration/test_kafka_bad_messages/test.py:
```diff
@@ -48,8 +94,9 @@ def test_bad_messages_parsing_stream(kafka_cluster):
         "MySQLDump",
     ]:
         print(format_name)
+        topic_name = f"{format_name}_{handle_error_mode}_err_{int(time.time())}"
```
Have you thought about just truncating the system table instead of using the timestamp as postfix? For me that feels more natural, but I don't have a strong preference, so feel free to keep it as is.

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue: changes per code review

ilejn commented:
I've changed 
" if (is_dead_letter && exception_message)" to "if (is_dead_letter) assert(exception_message) ..."
and switched to instance.query_with_retry .
Test failures are not related.
Awaiting for Marta's @morsapaes  philological advice ;)

antaljanosbenjamin reviewed (commented):
The changes looks good, the only question is the name. I asked @morsapaes to have a look at it, hopefully soon we will have an answer.

morsapaes commented:
Sorry, I missed the first ping. I think we should prioritize clarity over implementation details here, and `dead_letter_queue` clearly describes the purpose of the table. If you're wary of using `queue` when it isn't technically a queue (unlike _e.g._ `system.distribution_queue`) another option could be `dead_letter_errors`. I do find this name less intuitive, as an end user.

ilejn committed:
Revert "kafka_dead_letter_queue: rename dead_letter_queue to dead_letter"

This reverts commit 7ef4050cf928e2c38f3175b6e78a4b5b4c6b6a23.

antaljanosbenjamin reviewed (commented):
Apart from these small changes, the PR looks okay. Thanks for the nice work again!

ilejn committed:
kafka_dead_letter_queue: rename forgotten things

ilejn committed:
Merge remote-tracking branch 'origin/master' into kafka_dead_letter_queue

ilejn committed:
kafka_dead_letter_queue - rename - forgotten doc
